Adding basic coding guidelines/workflows for the NuGet contributor

nkolev92 opened the pull request:
## Bug

Fixes: https://github.com/NuGet/Home/issues/8991
Regression: No  
* Last working version:   
* How are we preventing it in future:   

## Fix

Details: 
All of the how to work with NuGet docs should be public and on GitHub!

We should avoid the wiki for workflow related things and use the repo itself.
It improves discoverability for contributors. It's in line with what the rest of the .NET repos do.
This is a first stab and what we are currently doing. 
The coding guidelines are a combination of https://github.com/NuGet/Home/wiki/Coding-Guidelines (very outdated), the .NET Framework coding guidelines, + any wofklow changes we have agreed to as team (for example, the usage of var was something @dtivel drove a few months back). 

This is not meant to be a controversial PR, merely a summary of the guidelines we are currently following/enforcing in our repo. 
If you think we should be making changes to our guidelines, this is not the PR you are looking for.

I will follow up this PR with some debugging/development docs, and potentially some arch docs later on. Specifically things that I think are missing, I will create issue for the respective tasks right before merging (I want to get some thoughts on what all should be in these docs)

I drew inspiration from dotnet repos such as project-system & runtime. 

- Debugging all 4 clients. 
- How do the tests work
- Build artifacts in details
- Basic arch (what is src/Nuget.Clients vs src/NuGet.Core etc), multi targeting etc.

fyi @aortiz-msft, the docs discussed during the team meeting. 

## Testing/Validation

Tests Added: No  
Reason for not adding tests: Not really test worthy :P :)   
Validation:

nkolev92 committed:
initial stab at coding guidelines

nkolev92 committed:
work

nkolev92 committed:
work

nkolev92 committed:
work

nkolev92 committed:
work

nkolev92 committed:
work

nkolev92 committed:
work

nkolev92 committed:
Add basic coding guidelines

zkat reviewed (changes requested):
I really like this! The one thing I'd really like to see is for `CONTRIBUTING.md` to be updated to point at these documents as appropriate, since GitHub integrates with that and the file has visibility where it matters. It might make sense to remove the "Getting Started guide" from `README.md` and replace it with a link to `CONTRIBUTING.md`.

zivkan commented on docs/coding-guidelines.md:
```diff
@@ -0,0 +1,403 @@
+
+1. We use four spaces of indentation (no tabs).
+
+1. We use `_camelCase` for internal and private fields and use `readonly` where possible. Prefix internal and private instance fields with `_`. Static fields are all CamelCase regardless of visibility. When used on static fields, `readonly` should come after `static` (e.g. `static readonly` not `readonly static`).  Public fields should be used sparingly and should use PascalCasing with no prefix  ...
```
```suggestion
1. We use `_camelCase` for internal and private fields and use `readonly` where possible. Prefix internal and private instance fields with `_`. Static fields are all PascalCase regardless of visibility. When used on static fields, `readonly` should come after `static` (e.g. `static readonly` not `readonly static`).  Public fields should be used sparingly and should use PascalCasing with no prefix when used.
```

zivkan commented on docs/coding-guidelines.md:
```diff
@@ -0,0 +1,403 @@
+   `namespace` declarations, and should be sorted alphabetically, with the exception of `System.*` namespaces, which are to be placed on top of all others.
+
+1. Avoid more than one empty line at any time. For example, do not have two
+   blank lines between members of a type.
```
new line in the middle of the sentence, and no whitespace before next list item.

zivkan commented on docs/coding-guidelines.md:
```diff
@@ -0,0 +1,403 @@
+   blank lines between members of a type.
+1. Avoid spurious free spaces.
+   For example avoid `if (someVar == 0)...`, where the dots mark the spurious free spaces.
+   Consider enabling "View White Space (Ctrl+R, Ctrl+W)" or "Edit -> Advanced -> View White Space" if using Visual Studio to aid detection.
```
When the markdown is rendered into HTML, the new lines get removed. Given earlier list items have multiple sentences on the same line, it's unclear if the intention here is to have one sentence per line, but in the same HTML paragraph, or if the intention was to have each sentence be rendered as a separate HTML paragraph.

nkolev92 replied:
I have it that way to improve readability of the doc.

nkolev92 replied:
changed it though for consistency.

zivkan commented on docs/coding-guidelines.md:
```diff
@@ -0,0 +1,403 @@
+   Consider enabling "View White Space (Ctrl+R, Ctrl+W)" or "Edit -> Advanced -> View White Space" if using Visual Studio to aid detection.
+
+1. If a file happens to differ in style from these guidelines (e.g. private members are named `m_member`
+   rather than `_member`), the existing style in that file takes precedence. Changes/refactorings are possible, but depending on the complexity, change frequency of the file, might need to be considered on their own merits in a separate pull request.
```
new line in middle of sentence.

zivkan commented on docs/coding-guidelines.md:
```diff
@@ -0,0 +1,403 @@
+        return string.IsNullOrEmpty(s)
+            ? null
+            : s.Trim();
+    }
```
I suggest removing the body of the method, as it doesn't add anything to the example (it doesn't contain any type names), and therefore is identical to the incorrect example, making it harder to see the differences.

nkolev92 replied:
It does contain type names: 

string.IsNullOrEmpty(s) vs
String.IsNullOrEmpty(s)

aortiz-msft commented on docs/getting-started.md:
```diff
@@ -0,0 +1,89 @@
+  - .NET Core Cross Platform Development
+  - .NET desktop development
+  - Visual Studio extension development.
+  - C++ development
```
Is this "Desktop development with C++"?

nkolev92 replied:
yep, fixed.

aortiz-msft commented on docs/getting-started.md:
```diff
@@ -0,0 +1,89 @@
+
+## Code
+
+Non NuGet members contribute to this repository is via the [fork model](https://help.github.com/articles/fork-a-repo/). Contributors push changes to their own "forked" version of NuGet.Client, and then submit a pull request into it requesting those changes be merged.
```
The way non-NuGet members....

zivkan commented on docs/coding-guidelines.md:
```diff
@@ -0,0 +1,403 @@
+
+When writing extension methods for an interface the sponsor type name must not start with an `I`.
+
+### Doc comments
```
Does this refer to XMLDoc, or docs.microsoft.com?

nkolev92 replied:
Fixed it to say Code comments.

zivkan reviewed (commented):
I think the coding-guidelines document should have a section near the top called "NuGet SDK" or something similar, saying that code under `src\NuGet.Core` and `src\NuGet.Client\NuGet.VisualStudio.Interop` is part of the NuGet SDK and breaking changes to public APIs will have a large impact on customers and therefore should be avoided. If APIs need to change, add overloads and mark the old API as obsolete, but do not remove. For these reasons, new internal and private APIs are preferred, to avoid growing the public APIs that need ongoing backwards compatibility.

nkolev92 committed:
work

nkolev92 commented:
Addressed all feedback. 

@zkat 

I pointed to the getting-started doc from the contributing.md.
My logic was to not have any doc pointers from root -> docs -> root. 
People that don't use github to navigate the docs, might have a harder time figuring out which information is where. 

Happy to chat more if necessary.

nkolev92 committed:
more of Kat's feedback

nkolev92 commented:
Made some more changes based on some offline feedback. 

README.md -> CONTRIBUTING.md -> everything in the docs repo, which is only workflow and coding guidelines right now.

zkat commented on CONTRIBUTING.md:
```diff
@@ -1,3 +1,98 @@
+## Resources for NuGet.Client development
+
+- [Workflow](workflow.md)
+- [Coding Guidelines](coding-guidelines.md)
```
these need to point to `docs/...`, right?

nkolev92 replied:
oops, yes it does. 
Fixing it.

zkat reviewed (approved):
just need to fix those links, and this LGTM! Thanks for putting these docs together!

nkolev92 committed:
Update CONTRIBUTING.md

nkolev92 committed:
Update CONTRIBUTING.md

nkolev92 committed:
add note that configure.ps1 and build.ps1 at least once is required

nkolev92 commented:
I have made some additional changes. 

@zivkan I agree with your comment, and actually I think the whole topic deserves its own section/topic and PR for that matter.

nkolev92 committed:
address feedback

dominoFire commented on CONTRIBUTING.md:
```diff
@@ -1,27 +1,100 @@
-Note that build.ps1 is the only process producing all nuget packages and artifacts, building from visual studio is good for development purposes, but is not enough at the moment.
-
-### Build.ps1 is responsible for the following steps:
+# Contributing
```
```suggestion
# NuGet.Client Contribution Guide
```

nkolev92 replied:
What's the benefit of that? 
Isn't it obvious?

dominoFire commented on README.md:
```diff
@@ -8,66 +8,23 @@ This repo contains the following clients:
+For how to contribute to this repo follow the [Contributing doc](CONTRIBUTING.md).
 
-### Steps to build NuGet client tools
+## Home repo
```
```suggestion
## NuGet/Home repo
```

nkolev92 replied:
What's the benefit? 
The names are relative the link is included.

nkolev92 replied:
I made them consistent throughout this doc.

dominoFire commented on README.md:
```diff
@@ -8,66 +8,23 @@ This repo contains the following clients:
 ## Feedback
 
-File bugs on [NuGet Home](https://github.com/nuget/home/issues).
+File NuGet.Client bugs on [NuGet Home](https://github.com/nuget/home/issues).
```
```suggestion
File NuGet.Client bugs on [NuGet/Home](https://github.com/nuget/home/issues).
```

nkolev92 replied:
Fixed.

dominoFire commented on docs/workflow.md:
```diff
@@ -0,0 +1,49 @@
+
+### Branching strategy
+
+The active development branch in our repo is `dev`. What we ship comes from the `release-majorminorx` branches.
```
```suggestion
The active development branch in our repo is `dev`. What we ship comes from the `release-<major>.<minor>` branches.
```

nkolev92 replied:
Fixed.

dominoFire commented on docs/workflow.md:
```diff
@@ -0,0 +1,49 @@
+The production source code is under the `src` folder.
+The test source code is under the `test` folder.
+The files affecting build are located under the `build` folder.
+The powershell files that are not part of the `Package Manager Console` are located under the `scripts` folder.
```
```suggestion
The PowerShell files that are not part of the `Package Manager Console` are located under the `scripts` folder. 
```

dominoFire commented on docs/coding-guidelines.md:
```diff
@@ -0,0 +1,402 @@
+
+Every repo also needs the Apache 2.0 License in a file called LICENSE.txt in the root of the repo. 
+
+## C# Coding Style
```
We need a section for NuGet Error codes

nkolev92 replied:
Feel like that goes beyond style, and it's more about you'd write a new feature etc. 
It's good suggestion, but my plan would be for that to get its own section.

dominoFire commented on docs/workflow.md:
```diff
@@ -0,0 +1,49 @@
+# Workflow
```
```suggestion
# NuGet.Client Development Workflow
```

nkolev92 replied:
Similar to the above, I don't see the point in blasting the repo name everywhere :)

dominoFire commented on docs/workflow.md:
```diff
@@ -0,0 +1,49 @@
+### Project naming pattern
+
+The general naming pattern is `NuGet.<area>.<subarea>`.
+
```
We should suggest add a description in the project file

nkolev92 replied:
Will add a new "how to add a new assembly section"

dominoFire commented on docs/workflow.md:
```diff
@@ -0,0 +1,49 @@
+The NuGet.Client repo currently has only one solution file named `NuGet.sln`. We do not want/need to have more than one solution file. 
+If deemed necessary by the team, we can consider solution filters at a future point.
+
+Every project in the NuGet.Client repo should be PackageReference based and if possible (read this as not .NET Framework WPF), an SDK based one. 
```
```suggestion
Every project in the NuGet.Client repo should be [PackageReference-based](https://docs.microsoft.com/en-us/nuget/consume-packages/package-references-in-project-files) and, if possible, an [SDK-styled](https://docs.microsoft.com/en-us/visualstudio/msbuild/how-to-use-project-sdk?view=vs-2019)  one.
```

dominoFire replied:
Consider using a bulleted list for the guidelines in this section

dominoFire commented on docs/workflow.md:
```diff
@@ -0,0 +1,49 @@
+
+The NuGet.Client repo currently has only one solution file named `NuGet.sln`. We do not want/need to have more than one solution file. 
+If deemed necessary by the team, we can consider solution filters at a future point.
+
```
Suggestions to add:

- Don't add references to external, non-NuGet packages, unless you have a strong justification for doing so.
- If you need a dependency to a NuGet library, prefer a `ProjectReference` over a `PackageReference` to that NuGet package
- Honor the target frameworks in `common.project.props`

nkolev92 replied:
> Don't add references to external, non-NuGet packages, unless you have a strong justification for doing so.

That's a good idea, I think it deserves its own section mentioning source-build etc.

nkolev92 replied:
All these are good ideas, but maybe not appropriate for this section. I will add them to a list of future improvements.

dominoFire commented on CONTRIBUTING.md:
```diff
@@ -1,27 +1,100 @@
+- Windows Powershell v3.0+
 
-Follow the instructions in the README.
+> Note that you can work on the NuGet.Client repo with Visual Studio 2017, but you will be unable to test the Visual Studio extension.
```
Is this still true now that we depend on MSBuild 16.4 ?

nkolev92 replied:
Yes.

dominoFire commented on docs/coding-guidelines.md:
````diff
@@ -0,0 +1,402 @@
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+```
+
+It is ok to skip it on generated files, such as `*.designer.cs`.
````
It is not mandatory to add the license header to Auto-generated files , such as `*.designer.cs`

dominoFire commented on docs/coding-guidelines.md:
```diff
@@ -0,0 +1,402 @@
+
+## Copyright header and license notice
+
+All source code files (mostly `src/**/*.cs` and `test/**/*.cs`) require this exact header (please do not make any changes to it):
```
```suggestion
All source code files (mostly `src/**/*.cs` and `test/**/*.cs`) require the _header listed below_ at the beginning of the file, without modifications:
```

dominoFire commented on docs/coding-guidelines.md:
````diff
@@ -0,0 +1,402 @@
+
+    ```cs
+    public string TrimString(string s) {
+        return string.IsNullOrEmpty(s)
````
```suggestion
    public string TrimString(string s)
    { 
        return string.IsNullOrEmpty(s)
```

nkolev92 replied:
Fixed

dominoFire commented on docs/coding-guidelines.md:
```diff
@@ -0,0 +1,402 @@
+By default all async methods must have the `Async` suffix. There are some exceptional circumstances where a method name from a previous framework will be grandfathered in.
+
+Passing cancellation tokens can be passed an optional parameter with a value of `default(CancellationToken)`, which is equivalent to `CancellationToken.None`.
+Prefer not having optional parameters.
```
Why?

nkolev92 replied:
See above. 

```
Optional parameters are allowed in certain situations, but discouraged.
Optional parameters are at the root for many versioning issues.
The optional parameter default values are a compile time decision, that's why adding a new optional parameter to an API is a breaking change!
Furthermore, if you have a method with optional parameters, adding an overload with additional optional parameters might cause a compile-time breaking change.
Optional parameters in internal/private methods are acceptable.
```

dominoFire commented on docs/coding-guidelines.md:
````diff
@@ -0,0 +1,402 @@
+// Assert
+Assert.AreEqual(1234, result);
+```
+
````
Is it possible to add a convention that each unit test should test only one part? i.e., don't make a giant unit test that test multiple parts of your code

nkolev92 replied:
What do you have in mind? 

Doesn't this cover it?
```
The crucial thing here is that the `Act` stage is exactly one statement. That one statement is nothing more than a call to the one method that you are trying to test. Keeping that one statement as simple as possible is also very important. 
```

dominoFire reviewed (commented):
Looks good. We need a section for NuGet Error Log codes.

nkolev92 committed:
address feedback

nkolev92 commented:
Thanks for the review @dominoFire 

I have addressed feedback where appropriate.
I think there are plenty of additional good suggestions, but I also think some of those need to be elaborated in their own docs (think sanity guide from one note). 

Here are some of my takeaways on new topics.

- NuGet and target frameworks

- NuGet + dependencies, how to add a new dependency in the NuGet client. Think source-build.

- All NuGet to NuGet dependencies are via a ProjectReference (implied via all NuGet assemblies version the same :) )

- How to add a new assembly section.

- We need a section for NuGet Error codes

I have added them to: https://github.com/NuGet/Home/issues/8991
